### PR TITLE
first proposal according #236

### PR DESCRIPTION
**Pull request by rasolca**
I refactored Cholesky according to Mikael suggestion.

new API (which dispatches to internal API):
```
dlaf::factorization::cholesky<backend, device, Type>(...)
```
new internal API:
```
template <Backend b, Device d, class T>
Cholesky{};

template <class T>
Cholesky<MC, CPU, T>{
// Note that call_L cannot be defined here, otherwise ETI doesn't work. 
static void call_L(..);
};
// non inline definition needed by ETI
template <class T>
void Cholesky<MC, CPU, T>::call_L(..){
  ..
}
```

The new directory structure is the following:
```
factorization/
    cholesky.h  <-  user API (template functions that dispatch to internal API)
    cholesky/
         cholesky.h  <-  internal template struct declaration
         cholesky_mc.h  <-  template partial specialization for MC backend
    # other algorithms e.g. QR
    qr.h
    qr/
         qr.h
         qr_mc.h
         t_factor.h
         t_factor_mc.h
```
 
@albestro @teonnik @msimberg What do you think?

### CODE REVIEW

**Comment by teonnik** [comment]
I like where this is heading. A few points of discussion:

The `cholesky/` contains the implementation details of the API. Can we rename things a little bit, for example:

`cholesky/cholesky.h` -> `cholesky/api.h` (i.e. the internal API of cholesky)
`cholesky/cholesky_mc.h` -> `cholesky/mc.h` (to avoid name repetitions)

File: include/dlaf/factorization/cholesky/mc.h [hunk 1]
```diff
@@ -22,18 +22,24 @@
...
-namespace mc {
 
-// Local implementation of Lower Cholesky factorization.
-template <class T>
-void cholesky_L(Matrix<T, Device::CPU>& mat_a) {
+template <Device device, class T>
+struct Cholesky<Backend::MC, device, T> {
+  static void call_L(Matrix<T, Device::CPU>& mat_a);
+  static void call_L(comm::CommunicatorGrid grid, Matrix<T, Device::CPU>& mat_a);
+};
```
Comment (teonnik) [inline include/dlaf/factorization/cholesky/mc.h:39]: Do we need the `Device device` template parameter here? It doesn't seem to be used in `call_L()`. Perhaps we can just set it to `Device::CPU` and remove it from `template<...>`

**Review by teonnik — CHANGES_REQUESTED** [review]

File: include/dlaf/factorization/factorization.h [hunk 2]
```diff
@@ -9,11 +9,4 @@
...
 
-#include "dlaf/types.h"
-
-namespace dlaf {
-
-template <Backend execution>
-struct Solver;
-
-}
+#include "dlaf/factorization/cholesky.h"
```
Comment (teonnik) [inline include/dlaf/factorization/factorization.h:12]: This header should probably replace the `factorization.h` header outside of the `factorization/` directory. That header has a wrong include statement in it (dlaf/factorization/mc.h), this one has the right include statement.

**Review by teonnik — APPROVED** [review]

**Comment by rasolca** [comment]
bors try

File: include/dlaf/auxiliary/norm/api.h [hunk 3]
```diff
@@ -12,8 +12,12 @@
 #include "dlaf/types.h"
 
 namespace dlaf {
+namespace auxiliary {
+namespace internal {
 
-template <Backend execution>
-struct Eigensolver;
+template <Backend backend, Device device, class T>
+struct Norm {};
```
Comment (albestro) [inline include/dlaf/auxiliary/norm/api.h:19]: like this.

**Review by albestro — APPROVED** [review]
Very good job!

Just a technical question: is there any benefit of using empty structs instead of just declaring them? See my comment just for reference in the code.

**Pull request merged**